Added DQM module for PPS dedicated random stream

AndreaBellora opened the pull request:
#### PR description:

This PR adds a DQM module meant to be run at T0-express with a new stream as input. 
The new express stream has been requested in https://its.cern.ch/jira/browse/CMSHLT-2606
The plots contain the number of digis per BX and plane in the PPS pixel detectors, which are necessary for monitoring the latency.
A backport to 13_0_0 will be submitted once the code is finalized.

#### PR validation:

Testing/validation of the module has been done running
`cmsRun DQM/CTPPS/test/pps_random_dqm_test_from_alcaraw_cfg.py` 
and using a test file produced following the instructions given in the HLT Jira ticket.

AndreaBellora committed:
Added DQM module for PPS random stream

AndreaBellora committed:
Code checks

AndreaBellora commented:
Adding @grzanka as watcher

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-40895/34378

- This PR adds an extra 16KB to repository

cmsbuild commented:
A new Pull Request was created by  @AndreaBellora  for master.

It involves the following packages:

- DQM/CTPPS (**dqm**)


@emanueleusai, @cmsbuild, @syuvivida, @rvenditti, @micsucmed, @pmandrik can you please review it and eventually sign? Thanks.
@fabferro, @grzanka this is something you requested to watch as well.
@perrotta, @dpiparo, @rappoccio you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

AndreaBellora commented:
There's actually still one pending point. This DQM module requires the PPS pixel raw2digi step to be performed before running the DQM source and harvester. 
This is done by cloning the equivalent alcareco step and adding it to the sequence in the test config:
```python
from EventFilter.CTPPSRawToDigi.ctppsRawToDigi_cff import ctppsPixelDigis as _ctppsPixelDigis
process.ctppsPixelDigisAlCaRecoProducer = _ctppsPixelDigis.clone(inputLabel = 'hltPPSCalibrationRaw')
```
Something similar will be needed in the sequence that runs at T0-express. Where should I define such a sequence or what is the best practice for this?

emanueleusai commented:
type ctpps

emanueleusai commented:
please test

cmsbuild commented:
-1

**Failed Tests**:   Build 
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-db5c7a/30961/summary.html
**COMMIT**: 432c76c5d9fbeafdc0f2ce86501ce5e275f6a9d7
**CMSSW**: CMSSW_13_1_X_2023-02-28-2300/el8_amd64_gcc11
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/40895/30961/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Build

I found compilation error when building: 

<pre>>> Compiling edm plugin /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_13_1_X_2023-02-28-2300/src/DQM/CTPPS/plugins/TotemT2DQMSource.cc
>> Compiling edm plugin /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_13_1_X_2023-02-28-2300/src/DQM/CTPPS/plugins/TotemTimingDQMSource.cc
>> Building edm plugin tmp/el8_amd64_gcc11/src/DQM/CTPPS/plugins/DQMCTPPSPlugins/libDQMCTPPSPlugins.so
/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02774/el8_amd64_gcc11/external/gcc/11.2.1-f9b9dfdd886f71cd63f5538223d8f161/bin/../lib/gcc/x86_64-redhat-linux-gnu/11.2.1/../../../../x86_64-redhat-linux-gnu/bin/ld: tmp/el8_amd64_gcc11/src/DQM/CTPPS/plugins/DQMCTPPSPlugins/ccGZQfoY.ltrans3.ltrans.o: in function `CTPPSRandomDQMSource::CTPPSRandomDQMSource(edm::ParameterSet const&)':
<artificial>:(.text+0x16ec): undefined reference to `vtable for CTPPSRandomDQMSource'
collect2: error: ld returned 1 exit status
gmake: *** [tmp/el8_amd64_gcc11/src/DQM/CTPPS/plugins/DQMCTPPSPlugins/libDQMCTPPSPlugins.so] Error 1
Leaving library rule at src/DQM/CTPPS/plugins
Entering library rule at DQM/CTPPS
>> Compiling  /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_13_1_X_2023-02-28-2300/src/DQM/CTPPS/src/TotemT2Segmentation.cc
>> Building shared library tmp/el8_amd64_gcc11/src/DQM/CTPPS/src/DQMCTPPS/libDQMCTPPS.so
</pre>

emanueleusai commented:
@AndreaBellora the code isn't compiling according to unit tests.
In the meantime we'll investigate how to add the sequence you need at T0-express

missirol commented on DQM/CTPPS/plugins/CTPPSRandomDQMSource.cc:
```diff
@@ -0,0 +1,201 @@
+//--------------------------------------------------------------------------
+
+void CTPPSRandomDQMSource::dqmBeginRun(edm::Run const &run, edm::EventSetup const &) {
+  for (int stn = 0; stn < StationIDMAX; stn++) {
+    StationStatus[stn] = 0;
+    for (int rp = 0; rp < RPotsIDMAX; rp++)
+      RPstatus[stn][rp] = 0;
+  }
+
+  unsigned int rpSts = rpStatusWord << 1;
+  for (int stn = 0; stn < NStationMAX; stn++) {
+    int stns = 0;
+    for (int rp = 0; rp < NRPotsMAX; rp++) {
+      rpSts = (rpSts >> 1);
+      RPstatus[stn][rp] = rpSts & 1;
+      if (RPstatus[stn][rp] > 0)
+        stns = 1;
+    }
+    StationStatus[stn] = stns;
+  }
+
+  for (int index = 0; index < 2 * 3 * NRPotsMAX; index++)
+    RPindexValid[index] = 0;
```
Nothing here depends on `run`, so it looks like it could be done in the constructor.

missirol commented on DQM/CTPPS/plugins/CTPPSRandomDQMSource.cc:
```diff
@@ -0,0 +1,201 @@
+//-------------------------------------------------------------------------------
+
+CTPPSRandomDQMSource::CTPPSRandomDQMSource(const edm::ParameterSet &ps)
+    : rpStatusWord(ps.getUntrackedParameter<unsigned int>("RPStatusWord", 0x8008)) {
+  tokenDigi = consumes<DetSetVector<CTPPSPixelDigi>>(ps.getUntrackedParameter<edm::InputTag>("tagRPixDigi"));
```
I don't think these should be `untracked`, as they affect the outputs of the DQM plugin.

perrotta replied:
> I don't think these should be `untracked`, as they affect the outputs of the DQM plugin.

@emanueleusai can correct me if I'm wrong, but as far as I remember there is the habit of keeping untracked the parameters in the DQM modules

missirol replied:
Okay, thanks. I didn't know this, so sorry for the confusion. It would be useful if someone (e.g. DQM) could clarify what rationale one should follow in this case (is it 'always untracked for DQM' ?).

AndreaBellora replied:
I'm leaving this pending. I will adjust it as soon as DQM clarifies.

emanueleusai replied:
right now it's a mix of both. Last year there was some initial work being done to uniform things and convert all params to untracked with the reasoning being that DQM does not produce "physics" quantities hence params should not be tracked. However when we started converting all modules to untracked, we left input collections as tracked for the reason Marino mentions above. 

In short, the current policy is: input collections tracked, everything else untracked

AndreaBellora replied:
Alright, so I think that the InputTag should be put as tracked, if we want to follow this convention. I'm doing it and pushing another commit.

missirol commented on DQM/CTPPS/plugins/CTPPSRandomDQMSource.cc:
```diff
@@ -0,0 +1,201 @@
+  int RPindexValid[RPotsTotalNumber];
+  MonitorElement *h2HitsVsBXRandoms[RPotsTotalNumber];
+
+  int getRPindex(int arm, int station, int rp) {
```
```suggestion
  int getRPindex(int arm, int station, int rp) const {
```
It does not modify any class member.

AndreaBellora replied:
Added.

missirol commented on DQM/CTPPS/plugins/CTPPSRandomDQMSource.cc:
```diff
@@ -0,0 +1,201 @@
+private:
+  edm::EDGetTokenT<edm::DetSetVector<CTPPSPixelDigi>> tokenDigi;
+
+  static constexpr int NArms = 2;
+  static constexpr int NStationMAX = 3;  // in an arm
+  static constexpr int NRPotsMAX = 6;    // per station
+  static constexpr int NplaneMAX = 6;    // per RPot
+  static constexpr int RPn_first = 3, RPn_last = 4;
+  static constexpr int StationIDMAX = 4;  // possible range of ID
+  static constexpr int RPotsIDMAX = 8;    // possible range of ID
```
```suggestion
  static constexpr int kNArms = 2;
  static constexpr int kNStationMAX = 3;  // in an arm
  static constexpr int kNRPotsMAX = 6;    // per station
  static constexpr int kNplaneMAX = 6;    // per RPot
  static constexpr int kRPn_first = 3, RPn_last = 4;
  static constexpr int kStationIDMAX = 4;  // possible range of ID
  static constexpr int kRPotsIDMAX = 8;    // possible range of ID
```

AndreaBellora replied:
Done.

missirol commented on DQM/CTPPS/plugins/CTPPSRandomDQMSource.cc:
```diff
@@ -0,0 +1,201 @@
+  CTPPSRandomDQMSource(const edm::ParameterSet &ps);
+  ~CTPPSRandomDQMSource() override;
+
+protected:
```
There is no `fillDescriptions` method. See for example
https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuideAboutPythonConfigFile#For_Component_developers

AndreaBellora replied:
Added it.

missirol commented on DQM/CTPPS/plugins/CTPPSRandomDQMSource.cc:
```diff
@@ -0,0 +1,201 @@
+
+  MonitorElement *hBX;
+
+  static constexpr int RPotsTotalNumber = NArms * NStationMAX * NRPotsMAX;
```
```suggestion
  static constexpr int kRPotsTotalNumber = NArms * NStationMAX * NRPotsMAX;
```

AndreaBellora replied:
Done.

missirol commented on DQM/CTPPS/plugins/CTPPSRandomDQMSource.cc:
```diff
@@ -0,0 +1,201 @@
+  static constexpr int StationIDMAX = 4;  // possible range of ID
+  static constexpr int RPotsIDMAX = 8;    // possible range of ID
+
+  unsigned int rpStatusWord = 0x8008;      // 220_fr_hr(stn2rp3)+ 210_fr_hr
+  int RPstatus[StationIDMAX][RPotsIDMAX];  // symmetric in both arms
+  int StationStatus[StationIDMAX];         // symmetric in both arms
+  const int IndexNotValid = 0;
```
Add trailing underscores to class members' names, e.g. `rpStatusWord_`.

Also, class members' names should start with lower-case letters

AndreaBellora replied:
Done, for all class members.

missirol commented on DQM/CTPPS/plugins/CTPPSRandomDQMSource.cc:
```diff
@@ -0,0 +1,201 @@
+
+//----------------------------------------------------------------------------------
+
+using namespace std;
+using namespace edm;
```
Maybe more of an opinion than a rule: avoid abusing `using namespace` when possible, simply prepend the appropriate namespace name where needed.

AndreaBellora replied:
I agree, I generally prefer avoiding the use of namespaces completely. I removed them.

missirol commented on DQM/CTPPS/plugins/CTPPSRandomDQMSource.cc:
```diff
@@ -0,0 +1,201 @@
+
+void CTPPSRandomDQMSource::bookHistograms(DQMStore::IBooker &ibooker, edm::Run const &, edm::EventSetup const &) {
+  ibooker.cd();
+  ibooker.setCurrentFolder("CTPPS/TrackingPixel");
```
Arguably, this should be passed via a configuration parameter in the PSet, e.g.
```
  ibooker.setCurrentFolder(folderName_);
  
  [..]
  
  process.mod = DQMEDAnalyzer('CTPPSRandomDQMSource',
    folderName = cms.string('CTPPS/TrackingPixel'),

  [..]
```

AndreaBellora replied:
Done, I took the opportunity to make the naming a bit clearer, as TrackingPixel is already used in the online DQM.

missirol commented on DQM/CTPPS/plugins/CTPPSRandomDQMSource.cc:
```diff
@@ -0,0 +1,201 @@
+//-------------------------------------------------------------------------------
+
+void CTPPSRandomDQMSource::analyze(edm::Event const &event, edm::EventSetup const &eventSetup) {
+  Handle<DetSetVector<CTPPSPixelDigi>> pixDigi;
+  event.getByToken(tokenDigi, pixDigi);
```
```suggestion
  auto const pixDigi = event.getHandle(tokenDigi_);
```

AndreaBellora replied:
Done.

missirol commented on DQM/CTPPS/plugins/CTPPSRandomDQMSource.cc:
```diff
@@ -0,0 +1,201 @@
+  Handle<DetSetVector<CTPPSPixelDigi>> pixDigi;
+  event.getByToken(tokenDigi, pixDigi);
+
+  cout << "Before checking digi" << endl;
```
`cout`s are not a good practise (and here they are even unguarded..). Use the standard CMSSW logging, e.g. `LogDebug` calls, see for example
https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuideMessageLogger

AndreaBellora replied:
Woops, my bad, that's not even useful, it was just for quick debugging. Removed!

missirol commented on DQM/CTPPS/plugins/CTPPSRandomDQMSource.cc:
```diff
@@ -0,0 +1,201 @@
+
+        ibooker.setCurrentFolder(rpd);
+
+        int indexP = getRPindex(arm, stn, rp);
+        RPindexValid[indexP] = 1;
+
+        h2HitsVsBXRandoms[indexP] = ibooker.book2D("Digi per plane per BX - random triggers",
```
Is `indexP` guaranteed to be inbounds? (i.e. smaller than the size of `h2HitsVsBXRandoms`)

AndreaBellora replied:
Yes, getRPindex allows for a maximum return value of 35, while kRPotsTotalNumber_ (which is the size of the mentioned arrays) is equal to 36. The two variables are linked, and the maximum adjusts depending on the other const variables which act as settings.

missirol commented on DQM/CTPPS/plugins/CTPPSRandomDQMSource.cc:
```diff
@@ -0,0 +1,201 @@
+
+        for (int p = 0; p < NplaneMAX; p++) {
+          CTPPSPixelDetId planeId(arm, stn, rp, p);
+          if ((*pixDigi).find(planeId.rawId()) != (*pixDigi).end()) {
+            int n_digis = (*pixDigi)[planeId.rawId()].size();
```
The second lookup could be avoided with something like this:
```suggestion
          auto pix_d = pixDigi->find(planeId.rawId());
          if (pix_d != pixDigi->end()) {
            int n_digis = pix_d->size();
```

AndreaBellora replied:
Indeed it's nicer. I'm adding this, thanks!

missirol reviewed (commented):
I'm not a DQM reviewer, but this PR caught my eye as it ignores many of the [CMSSW coding rules](https://cms-sw.github.io/cms_coding_rules.html). Up to DQM to enforce these suggestions if they find them useful.

AndreaBellora commented:
> @AndreaBellora the code isn't compiling according to unit tests. In the meantime we'll investigate how to add the sequence you need at T0-express

This is a bit weird, when I ran locally it passed the usual compilation/code checks. I'm checking again...

AndreaBellora commented:
@missirol Thank you very much! Actually, most of the code was inherited from DQM/CTPPS/plugins/CTPPSPixelDQMSource.cc, which I assumed was satisfying all coding rules. I'm implementing all the comments and pushing a fix

AndreaBellora commented on DQM/CTPPS/plugins/CTPPSRandomDQMSource.cc:
```diff
@@ -0,0 +1,201 @@
+class CTPPSRandomDQMSource : public DQMEDAnalyzer {
+public:
+  CTPPSRandomDQMSource(const edm::ParameterSet &ps);
+  ~CTPPSRandomDQMSource() override;
```
It did! Thanks!

AndreaBellora commented on DQM/CTPPS/plugins/CTPPSRandomDQMSource.cc:
```diff
@@ -0,0 +1,201 @@
+  void analyze(edm::Event const &e, edm::EventSetup const &eSetup) override;
+
+private:
+  edm::EDGetTokenT<edm::DetSetVector<CTPPSPixelDigi>> tokenDigi;
```
Done

missirol commented:
> I'm implementing all the comments and pushing a fix

@AndreaBellora , sounds good, thanks for addressing the comments.

AndreaBellora committed:
Added PR suggestions

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-40895/34394

- This PR adds an extra 20KB to repository

cmsbuild commented:
Pull request #40895 was updated. @emanueleusai, @cmsbuild, @syuvivida, @rvenditti, @micsucmed, @pmandrik can you please check and sign again.

emanueleusai commented:
please test

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-db5c7a/31029/summary.html
**COMMIT**: 3b7d1276b36ba19290a33322cc6d52fe7f4ae29f
**CMSSW**: CMSSW_13_1_X_2023-03-02-2300/el8_amd64_gcc11
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/40895/31029/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - You potentially added 8 lines to the logs
 - Reco comparison results: 13 differences found in the comparisons
 - DQMHistoTests: Total files compared: 49
 - DQMHistoTests: Total histograms compared: 3529699
 - DQMHistoTests: Total failures: 9
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3529668
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 48 files compared)
 - Checked 213 log files, 164 edm output root files, 49 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_13_1_X_2023-03-02-2300+db5c7a/55816/triggerResults): no differences found

AndreaBellora committed:
Making digi InputTag trackedd

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-40895/34431

- This PR adds an extra 16KB to repository

cmsbuild commented:
Pull request #40895 was updated. @emanueleusai, @cmsbuild, @syuvivida, @rvenditti, @micsucmed, @pmandrik can you please check and sign again.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-40895/34432

- This PR adds an extra 16KB to repository

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-40895/34433

- This PR adds an extra 16KB to repository

emanueleusai commented:
please test

emanueleusai commented:
works for us and we are ok with the names

AndreaBellora commented:
@francescobrivio thank you very much and sorry for the silence! However, are you sure that it would be enough? The input dataset contains only the PPS RAW collection (no digis), so a raw2digi step on it is needed. I defined it only in the test config:

https://github.com/cms-sw/cmssw/blob/159f060a58477758dc67f30e2efd770688dea3b9/DQM/CTPPS/test/pps_random_dqm_test_from_alcaraw_cfg.py#L47-L52

Shouldn't I add the raw2digi step in the 'source' sequence?
https://github.com/cms-sw/cmssw/blob/24d09a49a3ad130ed66cdb3452e0b304bc2386a6/DQMOffline/Configuration/python/DQMOffline_cff.py#L77

Anyway, there's a discussion ongoing to decide whether this will be put in the online DQM or offline. The decision will most likely be taken during tomorrow's meeting. As soon as this is clear I'll update the PR accordingly.

francescobrivio commented:
Ciao Andrea, I'm not very expert of PPS reconstruction and here I was only pointing out the missing steps to define a proper `@sequence` to be used later in Tier0, so yes you might need to add the `raw2digi` step somewhere in the sequence!
In case it's decided instead to go for the online client, these steps are probably not needed and you can define a complete sequence (with all the needed steps), directly in the client itself!

vavati commented:
As discussed at the DQM general meeting we wait for green light to proceed with the ONLINE solution. The sooner the better. Thanks

emanueleusai commented:
@AndreaBellora let me know when the code is ready for the "online" solution and I will review/approve.

AndreaBellora commented:
> @AndreaBellora let me know when the code is ready for the "online" solution and I will review/approve.

@emanueleusai apologies, I want to avoid misunderstandings, should I consider your reply as a greenlight for the online solution? I'm asking because in that case we also have to rename the input dataset to match the Online DQM scenario

AndreaBellora committed:
Added new DQM client for the PPSRandom stream

AndreaBellora committed:
Merge pull request #72 from CTPPS/CMSSW_13_0_0_pre4_RandomDQM_Online

Added new DQM client for the PPSRandom stream

AndreaBellora commented:
I added a client following the one used in the CTPPS online, could you provide guidance for testing it? Checking #35550 it seems that one was just tested with the equivalent of the test config I implemented here

cmsbuild commented:
Pull request #40895 was updated. @emanueleusai, @cmsbuild, @syuvivida, @rvenditti, @micsucmed, @pmandrik can you please check and sign again.

emanueleusai commented:
please test

emanueleusai commented:
@AndreaBellora we would test it in our playback machines. However I see a problem to do that right now, we need a backport to 13_0 to do that. Could you please prepare the backport? It will be needed anyway to deploy your code to DQM online machines.

AndreaBellora commented:
> @AndreaBellora we would test it in our playback machines. However I see a problem to do that right now, we need a backport to 13_0 to do that. Could you please prepare the backport? It will be needed anyway to deploy your code to DQM online machines.

Sure! The code should be 100% compatible with 13_0. I'll submit the PR this morning

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-db5c7a/31258/summary.html
**COMMIT**: 3f77d0055a0625aaa98c34d978a64109544fc5c3
**CMSSW**: CMSSW_13_1_X_2023-03-13-2300/el8_amd64_gcc11
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/40895/31258/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - You potentially added 24 lines to the logs
 - Reco comparison results: 7 differences found in the comparisons
 - DQMHistoTests: Total files compared: 49
 - DQMHistoTests: Total histograms compared: 3550756
 - DQMHistoTests: Total failures: 8
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3550726
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 48 files compared)
 - Checked 213 log files, 164 edm output root files, 49 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_13_1_X_2023-03-13-2300+db5c7a/55996/triggerResults): no differences found

AndreaBellora commented:
Backport created as PR #41049

emanueleusai commented:
hold

waiting for 13_0 streamers to test backport at P5

cmsbuild commented:
Pull request has been put on hold by  @emanueleusai
They need to issue an `unhold` command to remove the `hold` state or L1 can `unhold` it for all

vavati commented:
Streamers file available
/eos/cms/store/t0streamer/Data/ExpressPPSRandom/000/365/045
If i understood the rate was ~ 100Hz
ORM will ask to save these files, anyway for 1 week they should be safe

syuvivida commented:
@AndreaBellora @vavati 
I only got a chance to run a standalone CMSSW job with the client ppsrandom on these streamer files. The DQM output are here: /afs/cern.ch/work/s/syu/public/pps_output (this directory contains both root files and pb output)
Please check if the output is OK. If the output is fine, I could proceed to test using standard dqm test and maybe test the running with other clients in the same fu machines as well.

AndreaBellora commented:
@syuvivida Thank you very much! 
Indeed, I think that there might still be something funny in the folder structure of the output. If you could send me the instructions to reproduce your test it'd make the debugging easier.

syuvivida commented:
@AndreaBellora 
  I will send you instruction in a separate email.

AndreaBellora committed:
ROOT file directory renaming + stream name definition

cmsbuild commented:
Pull request #40895 was updated. @emanueleusai, @cmsbuild, @syuvivida, @rvenditti, @micsucmed, @pmandrik can you please check and sign again.

rappoccio commented:
Backport of #41049

emanueleusai commented:
please test

francescobrivio commented on DQM/Integration/python/clients/ppsrandom_dqm_sourceclient-live_cfg.py:
```diff
@@ -0,0 +1,97 @@
+    'keep FEDRawDataCollection_*_*_*'
+  )
+  
+process.source.streamLabel = "streamExpressPPSRandom"
```
From what I understood from the HLT report in the Daily Run Meetin (http://schwick.web.cern.ch/schwick/rctools/dailyReport/detailReport/display/2023-03-23) this should be updated to:
```suggestion
process.source.streamLabel = "streamDQMPPSRandom"
```

AndreaBellora replied:
Indeed! Thanks Francesco!

AndreaBellora committed:
Update DQM/Integration/python/clients/ppsrandom_dqm_sourceclient-live_cfg.py

Co-authored-by: francescobrivio <francesco.brivio@cern.ch>

francescobrivio commented:
Other two suggestions:
 1. you should add this to the unitTests suite:
 https://github.com/cms-sw/cmssw/blob/master/DQM/Integration/test/BuildFile.xml
 2. Eventually (once you have "real" streamer files with collisions), you could add the streamer files in `cms-data` under https://github.com/cms-data/DQM-Integration and update the unitTest to use `unittestinputsource_cfi`, similar to https://github.com/cms-sw/cmssw/pull/37667

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-40895/34797

- This PR adds an extra 16KB to repository

cmsbuild commented:
Pull request #40895 was updated. @emanueleusai, @cmsbuild, @syuvivida, @rvenditti, @micsucmed, @pmandrik can you please check and sign again.

vavati commented:
@francescobrivio : thanks. I guess we can do that once the beam back. Meantime I hope that tests can be done and put this stream in production. Right?

AndreaBellora committed:
Adding colz option

AndreaBellora committed:
Merge branch 'CMSSW_13_0_0_pre4_RandomDQM' of github.com:CTPPS/cmssw into CMSSW_13_0_0_pre4_RandomDQM

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-40895/34806

- This PR adds an extra 24KB to repository

cmsbuild commented:
Pull request #40895 was updated. @emanueleusai, @cmsbuild, @syuvivida, @rvenditti, @micsucmed, @pmandrik can you please check and sign again.

emanueleusai commented:
@francescobrivio @AndreaBellora yes let's keep the other two improvements for a separate PR

emanueleusai commented:
please test

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-db5c7a/31572/summary.html
**COMMIT**: a55d2c163a9227ec947c08d37fe33adb0485d008
**CMSSW**: CMSSW_13_1_X_2023-03-23-1100/el8_amd64_gcc11
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/40895/31572/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 6 differences found in the comparisons
 - DQMHistoTests: Total files compared: 49
 - DQMHistoTests: Total histograms compared: 3552750
 - DQMHistoTests: Total failures: 3
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3552725
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 48 files compared)
 - Checked 213 log files, 164 edm output root files, 49 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_13_1_X_2023-03-23-1100+db5c7a/56237/triggerResults): no differences found

emanueleusai commented:
+1

emanueleusai commented:
unhold

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @perrotta, @dpiparo, @rappoccio (and backports should be raised in the release meeting by the corresponding L2)

rappoccio commented:
+1
